papers/pydra/paper.rst

```diff
@@ -0,0 +1,965 @@
+element from the list.
+The results are grouped back when returning the result from the *Task*.
+While this example
+illustrates mapping and grouping of results over a single parameter, *Pydra*
+extends this to arbitrary combinations of input fields and downstream grouping
+over nested dataflows. Details of how splitters and combiners power *Pydra*'s
+scalable dataflows are described later.
+
+
+Submitter
+=========
+
+The ``Submitter`` class is responsible for unpacking *Workflows* and single
+*Tasks* with or without ``State`` into standalone stateless jobs that are then
+executed by *Workers*. When the *runnable* is a *Workflow*, the *Submitter* is
+responsible for checking if the *Tasks* from the graph are ready to run, i.e. if
+all the inputs are available, including the inputs that are set to the
+*Lazy Outputs* from previous *Tasks*. Once a *Task* is ready to run, the
+*Submitter* sends it to a *Worker*. When the runnable has a *State*, then the
+*Submitter* unpacks the *State* and sends multiple jobs to the *Worker* for the
+same *Task*. In order to avoid memory consumption as a result of scaling of *Tasks*,
+each job is sent as a pointer to a pickle file, together with information about
+its state, so that proper input can be retrieved just before running the *Task*.
+*Submitter* uses *AsyncIO* to manage all job executions to work in parallel,
+allowing scaling of execution as *Worker* resources are made available.
+
+Workers
+=======
+
+*Workers* in *Pydra* are responsible for the actual execution of the *Tasks* and
+are initialized by the *Submitter*. *Pydra* supports three types of execution
+managers: *ConcurrentFutures*, *Slurm* and *Dask* (experimental).
+When  ``ConcurrentFuturesWorker`` is created, ``ProcessPoolExecutor`` is used
+to create a "pool" for adding the runnables. ``SlurmWorker`` creates an`sbatch`
+submission script in order to execute the task, and ``DaskWorker`` make use of
+Dask's ``Client`` class and its ``submit`` method. All workers use
+*async functions* from *AsyncIO* in order to handle asynchronous processes. All
+*Workers* rely on a `load_and_run` function to execute each job from its pickled
+state.
+
+
+Key Features
+------------
+
+In this section, features of *Pydra* that exemplify its utility for scientific
+dataflows are presented. Individually, some of these features are present in the
+numerous workflow packages that exist, but Pydra is the only software that brings
+them together using a very lighweight codebase. The combination of the following
+features makes *Pydra* a powerful tool in scientific computation.
+
+Nested and Hashed Workflows
+===========================
+
+Scientific dataflows typically involve significant refinement and extensions as
+science and instrumentation evolves. *Pydra* was designed to provide an easy way
+of creating scientific dataflows that range from simple linear steps to complex
+nested graphs. It also enables reproducibility and reduced cost of dataflow
```

Comment:
Ja bym całe zdanie od "*Pydra* was designed to... " usunęła stąd. Rozprasza od technicznego aspektu o którym opowiadasz. Takie zdania albo do wstępu albo do zakończenia juz bym dawała.